src/ripple/app/misc/AMM_formulae.h

```diff
@@ -0,0 +1,253 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_AMM_FORMULAE_H_INCLUDED
+#define RIPPLE_APP_MISC_AMM_FORMULAE_H_INCLUDED
+
+#include <ripple/basics/IOUAmount.h>
+#include <ripple/basics/Number.h>
+#include <ripple/protocol/Issue.h>
+#include <ripple/protocol/Quality.h>
+#include <ripple/protocol/STAccount.h>
+#include <ripple/protocol/STAmount.h>
+
+#include <type_traits>
+
+namespace ripple {
+
+/** When converting Number to XRP as the result of swap in/out
+ * operation, round downward/upward respectively to maintain
+ * the invariant - the new pool product is greater or equal
+ * to the previous pool product.
+ */
+inline STAmount
+toSTAmount(
+    Issue const& issue,
+    Number const& n,
+    Number::rounding_mode mode = Number::rounding_mode::to_nearest)
+{
+    if (isXRP(issue))
+    {
+        Number::setround(mode);
+        auto const res = STAmount{issue, (std::int64_t)n};
+        Number::setround(Number::rounding_mode::to_nearest);
```

Comment:
This doesn't seem right - we unconditionally set the rounding mode to nearest regardless of the initial mode? Shouldn't this set the previous mode? Also, consider using a RAII-like class to save/restore the rounding mode.